base/abstractdict.jl

```diff
@@ -350,17 +370,24 @@ julia> ans == mergewith(+)(a, b)
 true
 ```
 """
-mergewith(combine, d::AbstractDict, others::AbstractDict...) =
+mergewith(combine, d::AbstractDict, others...) =
     mergewith!(combine, _typeddict(d, others...), others...)
 mergewith(combine) = (args...) -> mergewith(combine, args...)
-merge(combine::Callable, d::AbstractDict, others::AbstractDict...) =
+merge(combine::Callable, d::AbstractDict, others...) =
     merge!(combine, _typeddict(d, others...), others...)
 
 promoteK(K) = K
 promoteV(V) = V
-promoteK(K, d, ds...) = promoteK(promote_type(K, keytype(d)), ds...)
-promoteV(V, d, ds...) = promoteV(promote_type(V, valtype(d)), ds...)
-function _typeddict(d::AbstractDict, others::AbstractDict...)
+promoteK(K, d, ds...) = promoteK(promote_type(K, _keytype(d)), ds...)
+promoteV(V, d, ds...) = promoteV(promote_type(V, _valtype(d)), ds...)
+
+_keytype(d::AbstractDict) = keytype(d)
+_valtype(d::AbstractDict) = valtype(d)
+
+_keytype(d) = (p = eltype(d); p <: Pair ? fieldtype(p, 1) : Any)
+_valtype(d) = (p = eltype(d); p <: Pair ? fieldtype(p, 2) : Any)
```

Comment:
With `Any` here, we get:
```julia
julia> merge(Dict(1=>2), p for p in [3=>4.4])
Dict{Any, Any} with 2 entries:
  3 => 4.4
  1 => 2
```
A more clever algorithm could be copied from the `Dict` constructor in the future to get as nice as:
```julia
julia> merge(Dict(1=>2), Dict(p for p in [3=>4.4]))
Dict{Int64, Float64} with 2 entries:
  3 => 4.4
  1 => 2.0
```